Transclude widget: refresh selectively when needed

saqimtiaz opened the pull request:
Currently the `<$transclude>` widget calls `refreshSelf()` every time the transcluded tiddler changes. This means for instance that adding a tag to a tiddler results in every transclusion of its text field in the page template being re-rendered.

Instead `<$transclude/>` should only re-render when the source text or the parserType has changed.

saqimtiaz commented:
@Jermolene @pmario Would like to hear your thoughts on this when you get the chance.

Jermolene commented:
Hi @saqimtiaz that makes good sense, and definitely worth fixing.

saqimtiaz commented:
@Jermolene thank you for the quick feedback. I'll review the code in this PR in the coming week and see if there are improvements we can make in the implementation, and make sure all edge cases are covered.

Any thoughts you might have in terms of the code and implementation are very welcome.

pmario reviewed (commented):
Looks good to me

saqimtiaz commented on core/modules/widgets/transclude.js:
```diff
@@ -97,13 +99,36 @@ TranscludeWidget.prototype.makeRecursionMarker = function() {
+	
+TranscludeWidget.prototype.hasSourceTextChanged = function() {
+	var tiddler,
+			parserType,
```
@pmario as mentioned in the OP this is just some quick code I hacked together on a touchscreen to help with discussion, it needs clean up. 

Where feedback would be helpful is in the approach. The `hasSourceTextChanged` method has a lot in common with [wiki.parseTextReference](https://github.com/Jermolene/TiddlyWiki5/blob/8fbf52e419e71d726ea32b6c44e3ccfc4245d825/core/modules/wiki.js#L939) which the `transclude` widget calls during rendering. It would have been good to use the same code in both places to ensure consistency between them, but I am not sure there is an elegant and backwards compatible way to refactor `parseTextReference` to achieve this. Thoughts?

pmario replied:
I don't think, it has enough in common to change the `wiki.parseTextReference` ... and there are no tests for `wiki.parseTextReference` atm. So you'd need to create tests for all code paths first.

pmario commented on core/modules/widgets/transclude.js:
```diff
@@ -97,13 +99,36 @@ TranscludeWidget.prototype.makeRecursionMarker = function() {
+		tiddler = this.wiki.getTiddler(this.transcludeTitle);
+	}
+	if(tiddler) {
+		if(!this.transcludeIndex) {
```
As soon, as there is an "index" parameter, this code path will be executed. ... BUT the [widget docs](https://tiddlywiki.com/#TranscludeWidget) says, that the "field" parameter takes precedence, if it exists.

saqimtiaz commented:
@Jermolene this is now ready for review.

**Salient changes:**

1. `wiki.parseTextReference()` has been refactored. This method really did two things: (a) determine the correct text to parse and the parserType, and (b) ceate and return a parser if there was text to parse. 
   - The code for (a) has been refactored into a `wiki.getTextReferenceParserInfo()` method. This allows the same logic to be re-used from within the `<$transclude/>` widget which needs to know whether the source text or parser type has changed for better refresh handling.
   - I've added tests for `wiki.parseTextReference()` to ensure the behaviour is unchanged. 
2. In the `<$transclude>` widget I've added a `parserNeedsRefresh()` method which calls `wiki.getTextReferenceParserInfo()` and determines whether a refresh is needed. 
   - The new selective refresh behaviour depends on the parser having a `source` attribute. When the parser does not have this attribute - such as the HTML parser - the widget always refreshes when the `transcludeTitle` tiddler has changed, just as before. This allows us to separately add the source attribute to other parsers as desired for better refresh handling in `<$transclude/>`.


The way the diff presents the changes is a bit confusing, it may be easier to review the changed code in wiki.js directly:
https://github.com/Jermolene/TiddlyWiki5/blob/af5a24ba53453548eb849d49a491ef7a2406b018/core/modules/wiki.js#L939-L990

Note that the same end goal of selective refresh handling in `<$transclude/>` can be achieved without modifying `wiki.parseTextReference()`, however at the cost of some code duplication. I implemented it this way first (with extensive tests) and the code for it can be reviewed in the [first commit in this PR](https://github.com/Jermolene/TiddlyWiki5/pull/5736/commits/50a99339800ecf68147ae4b06ac1cfe6fafdb2e8). So if we wish to take that direction, all that is needed is for me to undo the last commit.

pmario commented:
>Note that the same end goal of selective refresh handling in <$transclude/> can be achieved without modifying wiki.parseTextReference(), however at the cost of some code duplication. I implemented it this way first (with extensive tests) and the code for it can be reviewed in the first commit in this PR. So if we wish to take that direction, all that is needed is for me to undo the last commit.

I'm OK with the new changes. ... At first it seemed, to be OK to duplicate some code, but then the new code seemed to be different to the documentation. ...

So I think you did choose the right path, with new tests and the refactoringl. IMO it will be more robust for future changes, since there is only 1 piece of code to take care of.

saqimtiaz commented:
I've improved the diff readability by changing the order of the methods wiki.parseTextReference() and wiki.getTextReferenceParserInfo in wiki.js

The code is otherwise unchanged.

Jermolene commented:
Thanks @saqimtiaz @pmario

BurningTreeC commented:
Hi @saqimtiaz, on the current prerelease page, if we open the preview pane of a tiddler then it's not possible to close it anymore. Is that a consequence of this PR (maybe) ?

saqimtiaz commented:
@BurningTreeC I will investigate in about 30 minutes.
